### PR TITLE
ThreatModal: Update flow

### PR DESCRIPTION
**Pull request by dkmyta**
## Description
Adds free and paid flows to the `ThreatModal`

### Other information:
- [ ] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion

- https://github.com/Automattic/jetpack-scan-team/issues/1548

## Does this pull request change what data or activity we track or use?

- No

## Testing instructions:

* Review `ThreatModal` storybooks

## Screenshots

Free:
Screen Shot on 2024-11-20 at 11-59-00

Paid:
- Default
Screen Shot on 2024-11-20 at 12-00-14

- Additional connections needed
Screen Shot 2024-11-20 at 12 01 24

- User connection needed
Screen Shot on 2024-11-20 at 12-01-36

- Site credentials needed
Screen Shot on 2024-11-20 at 12-01-58

### CODE REVIEW

**Comment by nateweller** [comment]
Screenshot 2024-11-21 at 1 50 17 PM

- [x] Focus ring touching the heading.
- [x] Possible to make the text itself clickable?

Screenshot 2024-11-21 at 1 51 30 PM

- [x] Inconsistent vertical spacing (probably due to `gap` and still rendering the details div at `0px` height)

Edit: [here is a PR](https://github.com/Automattic/jetpack/pull/40298) with some styling adjustments

File: projects/js-packages/components/components/threat-modal/fixer-state-notice.tsx [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+		title = __( 'The auto-fixer is taking longer than expected', 'jetpack' );
+		content = __(
+			'Jetpack has been attempting to auto-fix this threat for too long, and something may have gone wrong. Please try again later or contact support.',
+			'jetpack'
+		);
+	} else if ( fixerState.inProgress ) {
+		status = 'success';
+		title = __( 'An auto-fixer is in progress', 'jetpack' );
+		content = __( 'Please wait while Jetpack auto-fixes the threat.', 'jetpack' );
+	}
```
Comment (nateweller) [inline projects/js-packages/components/components/threat-modal/fixer-state-notice.tsx]:
```suggestion
	const { status, title, content } = useMemo( () => {
		if ( fixerState.error ) {
			return {
				status: 'error' as const,
				title: __( 'An error occurred auto-fixing this threat', 'jetpack' ),
				content: __(
					'Jetpack encountered a filesystem error while attempting to auto-fix this threat. Please try again later or contact support.',
					'jetpack'
				),
			};
		}

		if ( fixerState.stale ) {
			return {
				status: 'error' as const,
				title: __( 'The auto-fixer is taking longer than expected', 'jetpack' ),
				content: __(
					'Jetpack has been attempting to auto-fix this threat for too long, and something may have gone wrong. Please try again later or contact support.',
					'jetpack'
				),
			};
		}

		if ( fixerState.inProgress ) {
			return {
				status: 'success' as const,
				title: __( 'An auto-fixer is in progress', 'jetpack' ),
				content: __( 'Please wait while Jetpack auto-fixes the threat.', 'jetpack' ),
			};
		}

		return {};
	}, [ fixerState ] );
```
Nit

File: projects/js-packages/components/components/threat-modal/index.tsx [hunk 2]
```diff
@@ -65,88 +65,38 @@ export default function ThreatModal( {
...
-					handleConnectUser={ handleConnectUser }
+				<ThreatModalContext.Provider
+					value={ {
+						closeModal: modalProps.onRequestClose,
+					} }
 				>
-					<CredentialsGate
+					<ThreatFixConfirmation
+						threat={ threat }
+						fixerState={ fixerState }
```
Comment (nateweller) [inline projects/js-packages/components/components/threat-modal/index.tsx]: We're passing both the `threat` and `fixerState` into `ThreatFixConfirmation` - the `fixerState` is only used inside the component, and computed from the `threat`, so we could remove the computation of `fixerState` in this file and move it into `ThreatFixConfirmation` as necessary.

File: projects/js-packages/components/components/threat-modal/threat-fix-confirmation.tsx
Comment (nateweller) [inline projects/js-packages/components/components/threat-modal/threat-fix-confirmation.tsx:1]: What do you think about relying more on React Context to reduce the amount of prop drilling needed inside these modal sub-components?

File: projects/js-packages/components/components/threat-modal/threat-fix-confirmation.tsx [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+	handleUpgradeClick: () => void;
+	userConnectionNeeded: boolean;
+	userIsConnecting: boolean;
+	handleConnectUser: () => void;
+	siteCredentialsNeeded: boolean;
+	credentialsIsFetching: boolean;
+	credentialsRedirectUrl: string;
+	handleFixThreatClick?: ( threats: Threat[] ) => void;
+	handleIgnoreThreatClick?: ( threats: Threat[] ) => void;
+	handleUnignoreThreatClick?: ( threats: Threat[] ) => void;
```
Comment (nateweller) [inline projects/js-packages/components/components/threat-modal/threat-fix-confirmation.tsx]: General observation that formalizing our threat actions (callbacks, eligibility checks, etc) may simplify working on components like these. I have been meaning to take a run at it.

File: projects/js-packages/components/components/threat-modal/fixer-state-notice.tsx (hunk 1 above)
Comment (dkmyta) [reply projects/js-packages/components/components/threat-modal/fixer-state-notice.tsx]: Fixed in [220d28c](https://github.com/Automattic/jetpack/pull/40214/commits/220d28cb0b9495aaee61afe8a15f0c50735fa25e)

File: projects/js-packages/components/components/threat-modal/index.tsx (hunk 2 above)
Comment (dkmyta) [reply projects/js-packages/components/components/threat-modal/index.tsx]: Great point! Done in [94ed99a](https://github.com/Automattic/jetpack/pull/40214/commits/94ed99a29d2ae08197593f64b848df43d55e7102)

File: projects/js-packages/components/components/threat-modal/threat-fix-confirmation.tsx
Comment (dkmyta) [reply projects/js-packages/components/components/threat-modal/threat-fix-confirmation.tsx:1]: I agree fully! Are we thinking we should basically add all variables and methods for user connection, credentials and threat actions or only those that are reused in multiple places? Do we consider passing the main threat property?

Comment (dkmyta) [reply projects/js-packages/components/components/threat-modal/threat-fix-confirmation.tsx:1]: Took a pass at throwing everything into the context provider in [33268ee](https://github.com/Automattic/jetpack/pull/40214/commits/33268ee082fcfc3f07a36c2688bd9ed903174351) - works nicely and tidies things up beautifully.

File: projects/js-packages/components/components/threat-modal/threat-fix-confirmation.tsx (hunk 3 above)
Comment (dkmyta) [reply projects/js-packages/components/components/threat-modal/threat-fix-confirmation.tsx]: Are you referring to general organization of things? If so, I initially grouped things by the actions they are related to, but as that list grew things may have gotten out of order - open to any suggestions on how we could better do so.

File: projects/js-packages/components/components/threat-modal/styles.module.scss [hunk 4]
```diff
@@ -10,6 +10,24 @@
 	gap: calc( var( --spacing-base ) * 2 ); // 16px
 }
 
+.section .section__toggle {
+	text-decoration: none;
+
+	&:focus {
+		box-shadow: none;
+	}
```
Comment (nateweller) [inline projects/js-packages/components/components/threat-modal/styles.module.scss]: Need to maintain a focus style for accessibility purposes.

File: projects/js-packages/components/components/threat-modal/styles.module.scss [hunk 5]
```diff
@@ -10,6 +10,24 @@
 	gap: calc( var( --spacing-base ) * 2 ); // 16px
 }
 
+.section .section__toggle {
```
Comment (nateweller) [inline projects/js-packages/components/components/threat-modal/styles.module.scss:13]: Nit - Is the specificity necessary here? One of the benefits of the BEM style class naming is the ability to keep specificity flat (i.e. only one selector deep).

File: projects/js-packages/components/components/threat-modal/styles.module.scss [hunk 6]
```diff
@@ -10,6 +10,24 @@
...
+
+	&:focus {
+		box-shadow: none;
+	}
+
+	&:hover {
+		text-decoration: underline;
+	}
+
+	&_content {
```
Comment (nateweller) [inline projects/js-packages/components/components/threat-modal/styles.module.scss]: Nit/opinion - better to use full class names so they can be searched, i.e. ".section__toggle_content" vs "&_content".

Also, if "content" is a separate block, a double `__` might be wanted here.

**Review by nateweller — DISMISSED** [review]
Nice! Let's merge this :shipit: :+1:

**Pull request merged**